File: Examples/runtimes/python/DynamoDBEncryption/src/searchable_encryption/complex_example/beacon_config.py
# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
"""Sets up the beacon config."""
import boto3
from aws_cryptographic_material_providers.keystore.client import KeyStore
from aws_cryptographic_material_providers.keystore.config import KeyStoreConfig
from aws_cryptographic_material_providers.keystore.models import (
    KMSConfigurationKmsKeyArn,
)
from aws_cryptographic_material_providers.mpl.client import AwsCryptographicMaterialProviders
from aws_cryptographic_material_providers.mpl.config import MaterialProvidersConfig
from aws_cryptographic_material_providers.mpl.models import (
    CreateAwsKmsHierarchicalKeyringInput,
)
from aws_dbesdk_dynamodb.smithygenerated.aws_cryptography_dbencryptionsdk_dynamodb.models import (
    BeaconKeySourceSingle,
    BeaconVersion,
    CompoundBeacon,
    Constructor,
    ConstructorPart,
    DynamoDbTableEncryptionConfig,
    DynamoDbTablesEncryptionConfig,
    EncryptedPart,
    SearchConfig,
    SignedPart,
    SingleKeyStore,
    StandardBeacon,
)
from aws_dbesdk_dynamodb.smithygenerated.aws_cryptography_dbencryptionsdk_structuredencryption.models import (
    CryptoAction,
)


def setup_beacon_config(
    ddb_table_name: str,
    branch_key_id: str,
    branch_key_wrapping_kms_key_arn: str,
    branch_key_ddb_table_name: str,
):
    """Set up the beacon config."""
    keystore: KeyStore = KeyStore(
        KeyStoreConfig(
            ddb_table_name=branch_key_ddb_table_name,
            kms_configuration=KMSConfigurationKmsKeyArn(branch_key_wrapping_kms_key_arn),
            logical_key_store_name=branch_key_ddb_table_name,
            kms_client=boto3.client("kms"),
            ddb_client=boto3.client("dynamodb"),
        )
    )

    # Create standard beacons
    standard_beacon_list = [
        StandardBeacon(name="EmployeeID", length=4),
        StandardBeacon(name="TicketNumber", length=4),
        StandardBeacon(name="ProjectName", length=4),
        StandardBeacon(name="EmployeeEmail", length=4),
        StandardBeacon(name="CreatorEmail", length=4),
        StandardBeacon(name="ProjectStatus", length=4),
        StandardBeacon(name="OrganizerEmail", length=4),
        StandardBeacon(name="ManagerEmail", length=4),
        StandardBeacon(name="AssigneeEmail", length=4),
        StandardBeacon(name="City", loc="Location.City", length=4),
        StandardBeacon(name="Severity", length=4),
        StandardBeacon(name="Building", loc="Location.Building", length=4),
        StandardBeacon(name="Floor", loc="Location.Floor", length=4),
        StandardBeacon(name="Room", loc="Location.Room", length=4),
        StandardBeacon(name="Desk", loc="Location.Desk", length=4),
    ]

    # Define encrypted parts
    encrypted_part_list = [
        EncryptedPart(name="EmployeeID", prefix="E-"),
        EncryptedPart(name="TicketNumber", prefix="T-"),
        EncryptedPart(name="ProjectName", prefix="P-"),
        EncryptedPart(name="EmployeeEmail", prefix="EE-"),
        EncryptedPart(name="CreatorEmail", prefix="CE-"),
        EncryptedPart(name="ProjectStatus", prefix="PSts-"),
        EncryptedPart(name="OrganizerEmail", prefix="OE-"),
        EncryptedPart(name="ManagerEmail", prefix="ME-"),
        EncryptedPart(name="AssigneeEmail", prefix="AE-"),
        EncryptedPart(name="City", prefix="C-"),
        EncryptedPart(name="Severity", prefix="S-"),
        EncryptedPart(name="Building", prefix="B-"),
        EncryptedPart(name="Floor", prefix="F-"),
        EncryptedPart(name="Room", prefix="R-"),
        EncryptedPart(name="Desk", prefix="D-"),
    ]

    # Define signed parts
    signed_part_list = [
        SignedPart(name="TicketModTime", prefix="M-"),
        SignedPart(name="MeetingStart", prefix="MS-"),
        SignedPart(name="TimeCardStart", prefix="TC-"),
        SignedPart(name="ProjectStart", prefix="PS-"),
    ]

    employee_id_constructor_part = ConstructorPart(name="EmployeeID", required=True)
    ticket_number_constructor_part = ConstructorPart(name="TicketNumber", required=True)
    project_name_constructor_part = ConstructorPart(name="ProjectName", required=True)
    ticket_mod_time_constructor_part = ConstructorPart(name="TicketModTime", required=True)
    meeting_start_constructor_part = ConstructorPart(name="MeetingStart", required=True)
    time_card_start_constructor_part = ConstructorPart(name="TimeCardStart", required=True)
    employee_email_constructor_part = ConstructorPart(name="EmployeeEmail", required=True)
    creator_email_constructor_part = ConstructorPart(name="CreatorEmail", required=True)
    project_status_constructor_part = ConstructorPart(name="ProjectStatus", required=True)
    organizer_email_constructor_part = ConstructorPart(name="OrganizerEmail", required=True)
    project_start_constructor_part = ConstructorPart(name="ProjectStart", required=True)
    manager_email_constructor_part = ConstructorPart(name="ManagerEmail", required=True)
    assignee_email_constructor_part = ConstructorPart(name="AssigneeEmail", required=True)
    city_constructor_part = ConstructorPart(name="City", required=True)
    severity_constructor_part = ConstructorPart(name="Severity", required=True)
    building_constructor_part = ConstructorPart(name="Building", required=True)
    floor_constructor_part = ConstructorPart(name="Floor", required=True)
    room_constructor_part = ConstructorPart(name="Room", required=True)
    desk_constructor_part = ConstructorPart(name="Desk", required=True)

    # 6
    employee_id_constructor = Constructor(parts=[employee_id_constructor_part])
    ticket_number_constructor = Constructor(parts=[ticket_number_constructor_part])
    project_name_constructor = Constructor(parts=[project_name_constructor_part])
    ticket_mod_time_constructor = Constructor(parts=[ticket_mod_time_constructor_part])
    building_constructor = Constructor(parts=[building_constructor_part])

    meeting_start_floor_room_constructor = Constructor(
        parts=[meeting_start_constructor_part, floor_constructor_part, room_constructor_part]
    )

    time_card_start_employee_email_constructor = Constructor(
        parts=[time_card_start_constructor_part, employee_email_constructor_part]
    )

    time_card_start_constructor = Constructor(parts=[time_card_start_constructor_part])

    creator_email_constructor = Constructor(parts=[creator_email_constructor_part])

    project_status_constructor = Constructor(parts=[project_status_constructor_part])

    employee_email_constructor = Constructor(parts=[employee_email_constructor_part])

    organizer_email_constructor = Constructor(parts=[organizer_email_constructor_part])

    project_start_constructor = Constructor(parts=[project_start_constructor_part])

    manager_email_constructor = Constructor(parts=[manager_email_constructor_part])

    assignee_email_constructor = Constructor(parts=[assignee_email_constructor_part])

    city_constructor = Constructor(parts=[city_constructor_part])

    severity_constructor = Constructor(parts=[severity_constructor_part])

    building_floor_desk_constructor = Constructor(
        parts=[building_constructor_part, floor_constructor_part, desk_constructor_part]
    )

    # 7
    pk0_constructor_list = [
        employee_id_constructor,
        building_constructor,
        ticket_number_constructor,
        project_name_constructor,
    ]

    sk0_constructor_list = [
        ticket_mod_time_constructor,
        meeting_start_floor_room_constructor,
        time_card_start_employee_email_constructor,
        project_name_constructor,
        employee_id_constructor,
    ]

    pk1_constructor_list = [
        creator_email_constructor,
        employee_email_constructor,
        project_status_constructor,
        organizer_email_constructor,
    ]

    sk1_constructor_list = [
        meeting_start_floor_room_constructor,
        time_card_start_constructor,
        ticket_mod_time_constructor,
        project_start_constructor,
        employee_id_constructor,
    ]

    pk2_constructor_list = [manager_email_constructor, assignee_email_constructor]

    pk3_constructor_list = [city_constructor, severity_constructor]

    sk3_constructor_list = [building_floor_desk_constructor, ticket_mod_time_constructor]

    # 8
    compound_beacon_list = [
        CompoundBeacon(name="PK", split="~", constructors=pk0_constructor_list),
        CompoundBeacon(name="SK", split="~", constructors=sk0_constructor_list),
        CompoundBeacon(name="PK1", split="~", constructors=pk1_constructor_list),
        CompoundBeacon(name="SK1", split="~", constructors=sk1_constructor_list),
        CompoundBeacon(name="PK2", split="~", constructors=pk2_constructor_list),
        CompoundBeacon(name="PK3", split="~", constructors=pk3_constructor_list),
        CompoundBeacon(name="SK3", split="~", constructors=sk3_constructor_list),
    ]

    # 9
    beacon_versions = [
        BeaconVersion(
            standard_beacons=standard_beacon_list,
            compound_beacons=compound_beacon_list,
            encrypted_parts=encrypted_part_list,
            signed_parts=signed_part_list,
            version=1,  # MUST be 1
            key_store=keystore,
            key_source=BeaconKeySourceSingle(SingleKeyStore(key_id=branch_key_id, cache_ttl=6000)),
        )
    ]

    # 10. Create a Hierarchical Keyring
    mat_prov = AwsCryptographicMaterialProviders(MaterialProvidersConfig())

    keyring_input = CreateAwsKmsHierarchicalKeyringInput(
        branch_key_id=branch_key_id, key_store=keystore, ttl_seconds=6000
    )

    kms_keyring = mat_prov.create_aws_kms_hierarchical_keyring(keyring_input)

    # 11. Define crypto actions
    attribute_actions_on_encrypt = {
        # Our partition key must be configured as SIGN_ONLY
        "partition_key": CryptoAction.SIGN_ONLY,
        # Attributes used in beacons must be configured as ENCRYPT_AND_SIGN
        "EmployeeID": CryptoAction.ENCRYPT_AND_SIGN,
        "TicketNumber": CryptoAction.ENCRYPT_AND_SIGN,
        "ProjectName": CryptoAction.ENCRYPT_AND_SIGN,
        "EmployeeName": CryptoAction.ENCRYPT_AND_SIGN,
        "EmployeeEmail": CryptoAction.ENCRYPT_AND_SIGN,
        "CreatorEmail": CryptoAction.ENCRYPT_AND_SIGN,
        "ProjectStatus": CryptoAction.ENCRYPT_AND_SIGN,
        "OrganizerEmail": CryptoAction.ENCRYPT_AND_SIGN,
        "ManagerEmail": CryptoAction.ENCRYPT_AND_SIGN,
        "AssigneeEmail": CryptoAction.ENCRYPT_AND_SIGN,
        "City": CryptoAction.ENCRYPT_AND_SIGN,
        "Severity": CryptoAction.ENCRYPT_AND_SIGN,
        "Location": CryptoAction.ENCRYPT_AND_SIGN,
        # These are not beaconized attributes, but are sensitive data that must be encrypted
        "Attendees": CryptoAction.ENCRYPT_AND_SIGN,
        "Subject": CryptoAction.ENCRYPT_AND_SIGN,
        # Signed parts and unencrypted attributes can be configured as SIGN_ONLY or DO_NOTHING
        # For this example, we will set these to SIGN_ONLY to ensure authenticity
        "TicketModTime": CryptoAction.SIGN_ONLY,
        "MeetingStart": CryptoAction.SIGN_ONLY,
        "TimeCardStart": CryptoAction.SIGN_ONLY,
        "EmployeeTitle": CryptoAction.SIGN_ONLY,
        "Description": CryptoAction.SIGN_ONLY,
        "ProjectTarget": CryptoAction.SIGN_ONLY,
        "Hours": CryptoAction.SIGN_ONLY,
        "Role": CryptoAction.SIGN_ONLY,
        "Message": CryptoAction.SIGN_ONLY,
        "ProjectStart": CryptoAction.SIGN_ONLY,
        "Duration": CryptoAction.SIGN_ONLY,
    }

    # Define table encryption configuration
    table_configs = {}

    config = DynamoDbTableEncryptionConfig(
        logical_table_name=ddb_table_name,
        partition_key_name="partition_key",
        attribute_actions_on_encrypt=attribute_actions_on_encrypt,
        keyring=kms_keyring,
        search=SearchConfig(write_version=1, versions=beacon_versions),  # MUST be 1
    )

    # Store the configuration in a dictionary
    table_configs[ddb_table_name] = config

    # Return encryption configuration
    return DynamoDbTablesEncryptionConfig(table_encryption_configs=table_configs)

File: Examples/runtimes/go/searchableencryption/complexexample/beaconconfig.go
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package complexexample

import (
	"context"

	keystoreclient "github.com/aws/aws-cryptographic-material-providers-library/releases/go/mpl/awscryptographykeystoresmithygenerated"
	keystoretypes "github.com/aws/aws-cryptographic-material-providers-library/releases/go/mpl/awscryptographykeystoresmithygeneratedtypes"
	mpl "github.com/aws/aws-cryptographic-material-providers-library/releases/go/mpl/awscryptographymaterialproviderssmithygenerated"
	mpltypes "github.com/aws/aws-cryptographic-material-providers-library/releases/go/mpl/awscryptographymaterialproviderssmithygeneratedtypes"
	dbesdkdynamodbencryptiontypes "github.com/aws/aws-database-encryption-sdk-dynamodb/awscryptographydbencryptionsdkdynamodbsmithygeneratedtypes"
	dbesdkstructuredencryptiontypes "github.com/aws/aws-database-encryption-sdk-dynamodb/awscryptographydbencryptionsdkstructuredencryptionsmithygeneratedtypes"
	"github.com/aws/aws-database-encryption-sdk-dynamodb/dbesdkmiddleware"
	"github.com/aws/aws-database-encryption-sdk-dynamodb/examples/utils"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
	"github.com/aws/aws-sdk-go-v2/service/kms"
)

/*
 * This file is used in an example to demonstrate complex queries
 * you can perform using beacons.
 * The example data used is for demonstrative purposes only,
 * and might not meet the distribution and correlation uniqueness
 * recommendations for beacons.
 * See our documentation for whether beacons are
 * right for your particular data set:
 * https://docs.aws.amazon.com/database-encryption-sdk/latest/devguide/searchable-encryption.html#are-beacons-right-for-me
 *
 * This file sets up all the searchable encryption configuration required to execute the examples from
 * our workshop using the encryption client.
 */

func SetupBeaconConfig(
	ctx context.Context,
	ddbTableName,
	branchKeyId,
	branchKeyWrappingKmsKeyArn,
	branchKeyDdbTableName string) (*dynamodb.Client, error) {
	// 1. Create Keystore and branch key.
	//    These are the same constructions as in the Basic examples, which describe this in more detail.
	cfg, err := config.LoadDefaultConfig(ctx)
	if err != nil {
		return nil, err
	}

	kmsClient := kms.NewFromConfig(cfg)
	ddbClient := dynamodb.NewFromConfig(cfg)

	kmsConfig := keystoretypes.KMSConfigurationMemberkmsKeyArn{
		Value: branchKeyWrappingKmsKeyArn,
	}
	keyStoreConfig := keystoretypes.KeyStoreConfig{
		KmsClient:           kmsClient,
		DdbClient:           ddbClient,
		DdbTableName:        branchKeyDdbTableName,
		LogicalKeyStoreName: branchKeyDdbTableName,
		KmsConfiguration:    &kmsConfig,
	}

	keyStore, err := keystoreclient.NewClient(keyStoreConfig)
	if err != nil {
		return nil, err
	}

	// 2. Create standard beacons
	//    For this example, we use a standard beacon length of 4.
	//    The BasicSearchableEncryptionExample gives a more thorough consideration of beacon length.
	//    For production applications, one should always exercise rigor when deciding beacon length, including
	//    examining population size and considering performance.
	standardBeaconList := []dbesdkdynamodbencryptiontypes.StandardBeacon{
		{Name: "EmployeeID", Length: 4},
		{Name: "TicketNumber", Length: 4},
		{Name: "ProjectName", Length: 4},
		{Name: "EmployeeEmail", Length: 4},
		{Name: "CreatorEmail", Length: 4},
		{Name: "ProjectStatus", Length: 4},
		{Name: "OrganizerEmail", Length: 4},
		{Name: "ManagerEmail", Length: 4},
		{Name: "AssigneeEmail", Length: 4},
		{Name: "City", Loc: StringPtr("Location.City"), Length: 4},
		{Name: "Severity", Length: 4},
		{Name: "Building", Loc: StringPtr("Location.Building"), Length: 4},
		{Name: "Floor", Loc: StringPtr("Location.Floor"), Length: 4},
		{Name: "Room", Loc: StringPtr("Location.Room"), Length: 4},
		{Name: "Desk", Loc: StringPtr("Location.Desk"), Length: 4},
	}

	// 3. Define encrypted parts
	//    Note that some of the prefixes are modified from the suggested prefixes in Demo.md.
	//    This is because all prefixes must be unique in a configuration.
	//    Encrypted parts are described in more detail in the CompoundBeaconSearchableEncryptionExample.
	encryptedPartList := []dbesdkdynamodbencryptiontypes.EncryptedPart{
		{Name: "EmployeeID", Prefix: "E-"},
		{Name: "TicketNumber", Prefix: "T-"},
		{Name: "ProjectName", Prefix: "P-"},
		{Name: "EmployeeEmail", Prefix: "EE-"},
		{Name: "CreatorEmail", Prefix: "CE-"},
		{Name: "ProjectStatus", Prefix: "PSts-"},
		{Name: "OrganizerEmail", Prefix: "OE-"},
		{Name: "ManagerEmail", Prefix: "ME-"},
		{Name: "AssigneeEmail", Prefix: "AE-"},
		{Name: "City", Prefix: "C-"},
		{Name: "Severity", Prefix: "S-"},
		{Name: "Building", Prefix: "B-"},
		{Name: "Floor", Prefix: "F-"},
		{Name: "Room", Prefix: "R-"},
		{Name: "Desk", Prefix: "D-"},
	}

	// 4. Define signed parts.
	//    These are unencrypted attributes we would like to use in beacon queries.
	//    In this example, all of these represent dates or times.
	//    Keeping these attributes unencrypted allows us to use them in comparison-based queries. If a signed
	//    part is the first part in a compound beacon, then that part can be used in comparison for sorting.
	signedPartList := []dbesdkdynamodbencryptiontypes.SignedPart{
		{Name: "TicketModTime", Prefix: "M-"},
		{Name: "MeetingStart", Prefix: "MS-"},
		{Name: "TimeCardStart", Prefix: "TC-"},
		{Name: "ProjectStart", Prefix: "PS-"},
	}

	// 5. Create constructor parts.
	//    Constructor parts are used to assemble constructors (constructors described more in next step).
	//    For each attribute that will be used in a constructor, there must be a corresponding constructor part.
	//    A constructor part must receive:
	//     - name: Name of a standard beacon
	//     - required: Whether this attribute must be present in the item to match a constructor
	//    In this example, we will define each constructor part once and re-use it across multiple constructors.
	//    The parts below are defined by working backwards from the constructors in "PK Constructors",
	//        "SK constructors", etc. sections in Demo.md.
	employeeIdConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "EmployeeID", Required: true}
	ticketNumberConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "TicketNumber", Required: true}
	projectNameConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "ProjectName", Required: true}
	ticketModTimeConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "TicketModTime", Required: true}
	meetingStartConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "MeetingStart", Required: true}
	timeCardStartConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "TimeCardStart", Required: true}
	employeeEmailConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "EmployeeEmail", Required: true}
	creatorEmailConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "CreatorEmail", Required: true}
	projectStatusConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "ProjectStatus", Required: true}
	organizerEmailConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "OrganizerEmail", Required: true}
	projectStartConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "ProjectStart", Required: true}
	managerEmailConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "ManagerEmail", Required: true}
	assigneeEmailConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "AssigneeEmail", Required: true}
	cityConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "City", Required: true}
	severityConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "Severity", Required: true}
	buildingConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "Building", Required: true}
	floorConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "Floor", Required: true}
	roomConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "Room", Required: true}
	deskConstructorPart := &dbesdkdynamodbencryptiontypes.ConstructorPart{Name: "Desk", Required: true}

	// 6. Define constructors
	//    Constructors define how encrypted and signed parts are assembled into compound beacons.
	//    The constructors below are based off of the "PK Constructors", "SK constructors", etc. sections in Demo.md.

	// The employee ID constructor only requires an employee ID.
	// If an item has an attribute with name "EmployeeID", it will match this constructor.
	// If this is the first matching constructor in the constructor list (constructor list described more below),
	//     the compound beacon will use this constructor, and the compound beacon will be written as `E-X`.
	employeeIdConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*employeeIdConstructorPart}}
	ticketNumberConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*ticketNumberConstructorPart}}
	projectNameConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*projectNameConstructorPart}}
	ticketModTimeConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*ticketModTimeConstructorPart}}
	buildingConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*buildingConstructorPart}}

	// This constructor requires all of "MeetingStart", "Location.Floor", and "Location.Room" attributes.
	// If an item has all of these attributes, it will match this constructor.
	// If this is the first matching constructor in the constructor list (constructor list described more below),
	//     the compound beacon will use this constructor, and the compound beacon will be written as `MS-X~F-Y~R-Z`.
	// In a constructor with multiple constructor parts, the order the constructor parts are added to
	//     the constructor part list defines how the compound beacon is written.
	// We can rearrange the beacon parts by changing the order the constructors were added to the list.
	meetingStartFloorRoomConstructor := dbesdkdynamodbencryptiontypes.Constructor{
		Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*meetingStartConstructorPart, *floorConstructorPart, *roomConstructorPart},
	}

	timeCardStartEmployeeEmailConstructor := dbesdkdynamodbencryptiontypes.Constructor{
		Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*timeCardStartConstructorPart, *employeeEmailConstructorPart},
	}
	timeCardStartConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*timeCardStartConstructorPart}}
	creatorEmailConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*creatorEmailConstructorPart}}
	projectStatusConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*projectStatusConstructorPart}}
	employeeEmailConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*employeeEmailConstructorPart}}
	organizerEmailConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*organizerEmailConstructorPart}}
	projectStartConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*projectStartConstructorPart}}
	managerEmailConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*managerEmailConstructorPart}}
	assigneeEmailConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*assigneeEmailConstructorPart}}
	cityConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*cityConstructorPart}}
	severityConstructor := dbesdkdynamodbencryptiontypes.Constructor{Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*severityConstructorPart}}
	buildingFloorDeskConstructor := dbesdkdynamodbencryptiontypes.Constructor{
		Parts: []dbesdkdynamodbencryptiontypes.ConstructorPart{*buildingConstructorPart, *floorConstructorPart, *deskConstructorPart},
	}

	// 7. Add constructors to the compound beacon constructor list in desired construction order.
	//    In a compound beacon with multiple constructors, the order the constructors are added to
	//        the constructor list determines their priority.
	//    The first constructor added to a constructor list will be the first constructor that is executed.
	//    The client will evaluate constructors until one matches, and will use the first one that matches.
	//    If no constructors match, an attribute value is not written for that beacon.
	//    A general strategy is to add constructors with unique conditions at the beginning of the list,
	//       and add constructors with general conditions at the end of the list. This would allow a given
	//       item would trigger the constructor most specific to its attributes.
	pk0ConstructorList := []dbesdkdynamodbencryptiontypes.Constructor{
		employeeIdConstructor,
		buildingConstructor,
		ticketNumberConstructor,
		projectNameConstructor,
	}

	sk0ConstructorList := []dbesdkdynamodbencryptiontypes.Constructor{
		ticketModTimeConstructor,
		meetingStartFloorRoomConstructor,
		timeCardStartEmployeeEmailConstructor,
		projectNameConstructor,
		employeeIdConstructor,
	}

	pk1ConstructorList := []dbesdkdynamodbencryptiontypes.Constructor{
		creatorEmailConstructor,
		employeeEmailConstructor,
		projectStatusConstructor,
		organizerEmailConstructor,
	}

	sk1ConstructorList := []dbesdkdynamodbencryptiontypes.Constructor{
		meetingStartFloorRoomConstructor,
		timeCardStartConstructor,
		ticketModTimeConstructor,
		projectStartConstructor,
		employeeIdConstructor,
	}

	pk2ConstructorList := []dbesdkdynamodbencryptiontypes.Constructor{
		managerEmailConstructor,
		assigneeEmailConstructor,
	}

	pk3ConstructorList := []dbesdkdynamodbencryptiontypes.Constructor{
		cityConstructor,
		severityConstructor,
	}

	sk3ConstructorList := []dbesdkdynamodbencryptiontypes.Constructor{
		buildingFloorDeskConstructor,
		ticketModTimeConstructor,
	}

	// 8. Define compound beacons
	//    Compound beacon construction is defined in more detail in CompoundBeaconSearchableEncryptionExample.
	//    Note that the split character must be a character that is not used in any attribute value.
	compoundBeaconList := []dbesdkdynamodbencryptiontypes.CompoundBeacon{
		{Name: "PK", Split: "~", Constructors: pk0ConstructorList},
		{Name: "SK", Split: "~", Constructors: sk0ConstructorList},
		{Name: "PK1", Split: "~", Constructors: pk1ConstructorList},
		{Name: "SK1", Split: "~", Constructors: sk1ConstructorList},
		{Name: "PK2", Split: "~", Constructors: pk2ConstructorList},
		{Name: "PK3", Split: "~", Constructors: pk3ConstructorList},
		{Name: "SK3", Split: "~", Constructors: sk3ConstructorList},
	}

	// 9. Create BeaconVersion.
	beaconVersion := &dbesdkdynamodbencryptiontypes.BeaconVersion{
		StandardBeacons: standardBeaconList,
		CompoundBeacons: compoundBeaconList,
		EncryptedParts:  encryptedPartList,
		SignedParts:     signedPartList,
		Version:         1, // MUST be 1
		KeyStore:        keyStore,
		KeySource: &dbesdkdynamodbencryptiontypes.BeaconKeySourceMembersingle{
			Value: dbesdkdynamodbencryptiontypes.SingleKeyStore{
				KeyId:    branchKeyId,
				CacheTTL: 6000,
			},
		},
	}

	beaconVersions := []dbesdkdynamodbencryptiontypes.BeaconVersion{*beaconVersion}

	// 10. Create a Hierarchical Keyring
	matProv, err := mpl.NewClient(mpltypes.MaterialProvidersConfig{})
	utils.HandleError(err)

	hierarchicalKeyringInput := mpltypes.CreateAwsKmsHierarchicalKeyringInput{
		KeyStore:    keyStore,
		BranchKeyId: &branchKeyId,
		TtlSeconds:  6000, // This dictates how often we call back to KMS to authorize use of the branch keys
	}

	kmsKeyring, err := matProv.CreateAwsKmsHierarchicalKeyring(context.Background(), hierarchicalKeyringInput)
	utils.HandleError(err)

	// 11. Define crypto actions
	attributeActionsOnEncrypt := map[string]dbesdkstructuredencryptiontypes.CryptoAction{
		// Our partition key must be configured as SIGN_ONLY
		"partition_key": dbesdkstructuredencryptiontypes.CryptoActionSignOnly,
		// Attributes used in beacons must be configured as ENCRYPT_AND_SIGN
		"EmployeeID":     dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"TicketNumber":   dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"ProjectName":    dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"EmployeeName":   dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"EmployeeEmail":  dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"CreatorEmail":   dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"ProjectStatus":  dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"OrganizerEmail": dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"ManagerEmail":   dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"AssigneeEmail":  dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"City":           dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"Severity":       dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"Location":       dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		// These are not beaconized attributes, but are sensitive data that must be encrypted
		"Attendees": dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		"Subject":   dbesdkstructuredencryptiontypes.CryptoActionEncryptAndSign,
		// signed parts and unencrypted attributes can be configured as SIGN_ONLY or DO_NOTHING
		// For this example, we will set these to SIGN_ONLY to ensure authenticity
		"TicketModTime": dbesdkstructuredencryptiontypes.CryptoActionSignOnly,
		"MeetingStart":  dbesdkstructuredencryptiontypes.CryptoActionSignOnly,
		"TimeCardStart": dbesdkstructuredencryptiontypes.CryptoActionSignOnly,
		"EmployeeTitle": dbesdkstructuredencryptiontypes.CryptoActionSignOnly,
		"Description":   dbesdkstructuredencryptiontypes.CryptoActionSignOnly,
		"ProjectTarget": dbesdkstructuredencryptiontypes.CryptoActionSignOnly,
		"Hours":         dbesdkstructuredencryptiontypes.CryptoActionSignOnly,
		"Role":          dbesdkstructuredencryptiontypes.CryptoActionSignOnly,
		"Message":       dbesdkstructuredencryptiontypes.CryptoActionSignOnly,
		"ProjectStart":  dbesdkstructuredencryptiontypes.CryptoActionSignOnly,
		"Duration":      dbesdkstructuredencryptiontypes.CryptoActionSignOnly,
	}

	// 12. Set up table config
	tableConfig := &dbesdkdynamodbencryptiontypes.DynamoDbTableEncryptionConfig{
		LogicalTableName:          ddbTableName,
		PartitionKeyName:          "partition_key",
		AttributeActionsOnEncrypt: attributeActionsOnEncrypt,
		Keyring:                   kmsKeyring,
		Search: &dbesdkdynamodbencryptiontypes.SearchConfig{
			WriteVersion: 1, // MUST be 1
			Versions:     beaconVersions,
		},
	}

	tableConfigs := map[string]dbesdkdynamodbencryptiontypes.DynamoDbTableEncryptionConfig{
		ddbTableName: *tableConfig,
	}

	// 13. Create the DynamoDb Encryption Interceptor
	encryptionConfig := dbesdkdynamodbencryptiontypes.DynamoDbTablesEncryptionConfig{
		TableEncryptionConfigs: tableConfigs,
	}

	// 14. Create a new AWS SDK DynamoDb client using the DynamoDb Encryption Interceptor above
	dbEsdkMiddleware, err := dbesdkmiddleware.NewDBEsdkMiddleware(encryptionConfig)
	utils.HandleError(err)
	ddb := dynamodb.NewFromConfig(cfg, dbEsdkMiddleware.CreateMiddleware())

	return ddb, nil
}
